test commodity interfaces

bereba opened the pull request:
Added script and input data to check commodity interfaces between model structure of SEDOS sectors.
Please check the script and the location of its input data, which basically are the process_set from the model structure excel.

bereba committed:
Added script and input data to check commodity interfaces between model structure of SEDOS sectors

FelixMau committed:
Merge branch 'dev' into feature-check-commodity-interfaces

FelixMau committed:
Check commodity interfaces

Using Structure
Added visual representation by Beneharo to Structure class

Added `only suspicious` candidates to Structure class

FelixMau committed:
Linting

FelixMau committed:
Adding Comments

FelixMau committed:
Try to reduce complexity

FelixMau commented:
I updated your code to use the Structures class so we can add the functionality to the adapter directly. I hope you don't mind @bereba

henhuy commented on data_adapter/structure.py:
```diff
@@ -150,3 +152,149 @@ def _init_parameters(self, parameter_sheet: str) -> dict:
+        # Save the plot as svg
+        plt.show()
+
+    def get_commodity_diff(self, input_processes: list = ["source", "import"], output_processes: list = ["sink"]):
```
One question: Do we need to set input and output processes explicitly? I would think, that processes which either have only inputs or only outputs are "creating" or "destroying" commodities implicitly - that would it make easier in case some processes are not named sink/source or import...

FelixMau replied:
Haven't thought it that way, thank you!

bereba replied:
Maybe it would also make sense to have an additional color (e.g. blue) for commodities that only come from  imports/sources. I think for the interface check this might be an interesting value.

sinks and exports could be filtered as mentioned above. However, sources and imports should be checked. I accordingly adjust my proposed filter in the review below.

FelixMau replied:
I have added blue for commodities that are only outputs of processes (but don't go anywhere) and purple for commodities that have no source to them (picture below) 
![Figure_1](https://github.com/sedos-project/data_adapter/assets/98911066/bad3f243-3ff1-4f84-80a6-723313425031)

FelixMau replied:
This function is also separated from the function for drawing the commodities. 
It is meant only to get a list of "open" commodities, therefore I would like to include sources and sinks (in opposite of my thinking above) to see what commodities are either not destroyed or not created yet.

bereba replied:
I am not sure if this really helps, as now all these open commodities are blue and we do not know if it is an input or output, right? We could see those processes before, if one line did only have green or red cell.
The idea of the additional color was more to highlight commodities which have no other upstream processes except the imports/sources.

If you want it more clear that there is something "open" you could maybe use another formating. Like highighting the text or something?

FelixMau replied:
Yes we did know what inputs/outputs were but I did not understand correctly what you wanted :)
So here is the improved version for the Steel Industry Part O1: 
![Figure_1](https://github.com/sedos-project/data_adapter/assets/98911066/9da67248-b5fe-4abd-abf6-d0572944a6b3)

henhuy commented on tests/test_commodity_interfaces.py:
This is not a real test.
Better option would be to add a minimal structure with predefined missing sinks or sources and write a test which checks if missing parts are found. If you don't have the time to do so, I will merge nevertheless as this function is optional.

henhuy reviewed (dismissed):
Only checked if its working - did not reviewed code itself.
Have two recommendations to think about.
But you can merge anyway if you want!

henhuy commented:
Made matplotlib optional and moved structure file to `test_data/test_structures`

henhuy commented:
Also linter is not happy with `structure.py` due to your changes...

henhuy committed:
Make matplotlib optional

FelixMau committed:
Turn into real testing

FelixMau committed:
Remove useless imports

FelixMau committed:
Happy Linter is good Linter

FelixMau commented:
Separated parts of the plot function outside the class to make the linter happy.

FelixMau committed:
Keep local functions local

bereba commented on tests/test_commodity_interfaces.py:
```diff
@@ -2,100 +2,18 @@
-df = df[~df['process'].str.contains('storage|transport|import|export')]
-
-# Delete square brackets from dataframe
-df['input'] = df['input'].str.replace(r'[][]', '', regex=True)
```
In my last review I additionally deleted blank spaces to avoid unintended double naming and the _orig ending for commodities to simplify the graph.

bereba replied:
df['input'] = df['input'].str.replace(r' ', '', regex=True)
df['output'] = df['output'].str.replace(r' ', '', regex=True)
df['input'] = df['input'].str.replace(r'_orig', '', regex=True)
df['output'] = df['output'].str.replace(r'_orig', '', regex=True)

FelixMau replied:
I think it might be even necessary to see the blank spaces. Since I am using the data_adapter itself we have to have a consistent naming here and might miss an opportunity to identify potential typos.

bereba replied:
I don't understand this point. Could you elaborate this more please?
We are talking about the naming of inputs and outputs, which might be separated with commas, right?

FelixMau replied:
Yes but the whitespace trimming already happens within the data_adapter beforehand, as the plot function is now part of the class I thought it is better to reuse already existing parts of code 
To improve consistency I want to use exactly what is going to be input for the data_adapter_oemof.

bereba replied:
Ah okay now I understand. Makes sense.

bereba commented on tests/test_commodity_interfaces.py:
```diff
@@ -2,100 +2,18 @@
+structure.plot_commodity_interfaces()
 
-# Filter out processes which impair the logic
-df = df[~df['process'].str.contains('storage|transport|import|export')]
```
In my last interface check I also filtered the following key words (due to changes in the model structure list):
storage|export|helper_source|helper_sink'

bereba replied:
As elaborated in henhuys comment my suggestion for the filters is as follows:

'storage|export||helper_sink|delivery'

As "_orig" could neglected I would also add the delivery processes.
What do you think?

FelixMau replied:
As I am not very deeply involved into discussions about naming and different edge cases, I don't really understand what you want to achieve with the filtering and what effect the exact process naming has. 
I don't understand what benefit comes not including sinks and sources in the plot?

I do understand that excluding `storage` makes sense to make the plot smaller. Even though it can have multiple outputs/inputs as  `helper_co2_storage_permanent` that has an output going into "nowhere" and for modelling I eventually need to be aware of such issues to help debugging.

bereba replied:
The idea of this plot first was mainly to check if there are defined inputs which are not being created or outputs which are not being used.

Therefore, temporal (storages) or spatial (delivery) shifts are disturbing this logic, as it might appear that e.g. electricity is present in our system as input and output, although there is no process to generate it.

For the sinks and exports, it is a modeling-specific decision I guess, and if it helps you, you can also keep them.

FelixMau replied:
Ah yes now I see, thank you!

So delivery should only be a kind of transport? As in the Model structure unfortunately there are processes that seem to be also doing Conversion. Such as `helper_co2_delivery`

bereba replied:
Yes, as we have a one-node model approach, the delivery is always modeled with a conversion between commodities.
For helper_co2_delivery different commodities are additionally merged.

bereba reviewed (commented):
I recently changed some details for the interface check in the code locally.
You might want to apply this in your new functions as well.

FelixMau committed:
Making origin visible

Sources are now allocateable to source sector
Added filter to skip:
storages, delivery and export

FelixMau committed:
Linternator

FelixMau committed:
Remove plot from test

FelixMau committed:
Bugremoval

henhuy commented:
Is this ready to merge?

FelixMau commented:
From my side, yes
